pkg/syncer/upsync/upsync_process_test.go

```diff
@@ -0,0 +1,762 @@
+	return map[schema.GroupVersionResource]ddsif.GVRPartialMetadata{
+		{
+			Version:  "v1",
+			Resource: "namespaces",
+		}: {
+			Scope: apiextensionsv1.ClusterScoped,
+			Names: apiextensionsv1.CustomResourceDefinitionNames{
+				Singular: "namespace",
+				Kind:     "Namespace",
+			},
+		},
+		{
+			Version:  "v1",
+			Resource: "persistentvolumes",
+		}: {
+			Scope: apiextensionsv1.ClusterScoped,
+			Names: apiextensionsv1.CustomResourceDefinitionNames{
+				Singular: "persistentvolume",
+				Kind:     "PersistentVolume",
+			},
+		},
+		{
+			Version:  "v1",
+			Resource: "pods",
+		}: {
+			Scope: apiextensionsv1.NamespaceScoped,
+			Names: apiextensionsv1.CustomResourceDefinitionNames{
+				Singular: "pod",
+				Kind:     "Pod",
+			},
+		},
+	}
+}
+
+func (s *mockedGVRSource) Ready() bool {
+	return true
+}
+
+func (s *mockedGVRSource) Subscribe() <-chan struct{} {
+	return make(<-chan struct{})
+}
+
+func TestUpsyncerprocess(t *testing.T) {
+	type testCase struct {
+		fromNamespace   *corev1.Namespace
+		toNamespaceName string
+		gvr             schema.GroupVersionResource
+		fromResource    runtime.Object
+		toResource      runtime.Object
+		doOnDownstream  func(tc testCase, client dynamic.Interface)
+
+		resourceToProcessName string
+
+		upstreamLogicalCluster logicalcluster.Name
+		syncTargetName         string
+		syncTargetClusterName  logicalcluster.Name
+		syncTargetUID          types.UID
+		expectError            bool
+		expectRequeue          bool
+		expectActionsOnFrom    []clienttesting.Action
+		expectActionsOnTo      []kcptesting.Action
+		includeStatus          bool
+	}
+	tests := map[string]testCase{
+		"Upsyncer upsyncs namespaced resources": {
+			upstreamLogicalCluster: "root:org:ws",
+			toNamespaceName:        "test",
+			fromNamespace: namespace("kcp-33jbiactwhg0", "", map[string]string{
+				"internal.workload.kcp.io/cluster": "6ohB8yeXhwqTQVuBzJRgqcRJTpRjX7yTZu5g5g",
+			},
+				map[string]string{
+					"kcp.io/namespace-locator": `{"syncTarget": {"cluster":"root:org:ws", "name":"us-west1", "uid":"syncTargetUID"}, "cluster":"root:org:ws","namespace":"test"}`,
+				},
+			),
+			gvr: schema.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"},
+			fromResource: createPod("test-pod", "kcp-33jbiactwhg0", "", map[string]string{
+				"internal.workload.kcp.io/cluster": "6ohB8yeXhwqTQVuBzJRgqcRJTpRjX7yTZu5g5g",
+				workloadv1alpha1.ClusterResourceStateLabelPrefix + "6ohB8yeXhwqTQVuBzJRgqcRJTpRjX7yTZu5g5g": "Upsync",
+			}, nil, nil, "1"),
+			toResource:            nil,
+			resourceToProcessName: "test-pod",
+			syncTargetName:        "us-west1",
+			expectActionsOnFrom:   []clienttesting.Action{},
+			expectActionsOnTo: []kcptesting.Action{
+				kcptesting.NewCreateAction(schema.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"}, logicalcluster.NewPath("root:org:ws"), "test", toUnstructured(t, createPod("test-pod", "test", "", map[string]string{
```

Comment:
a `unmarshalYaml(t, ```...```)` would make this much easier to read.

Reply:
I made a try, but I'm not convinced it makes it so better.
Could we keep that for a distinct PR where we would would change the unit tests of all the controller: Upsyncer, SpecSyncer, StatsSyncer, which all follow the same structure ?

Reply:
After our dicsussion, I fixed it by refactoring the test to use a small dsl, in order to both increase readability while avoiding more verbosity: see commit 80a228ba11dc7c931ecb334c327e8d9fc7653278